otter/models/cass.py

```diff
@@ -1831,43 +1826,53 @@ def _dict(r): return json.loads(r['server_blob'])
 
         yield do_return((list(rfunc(rows)), last_update))
 
-    def insert_servers(self, last_update, servers, clear_others):
-        """
-        See :method:`IScalingGroupServersCache.insert_servers`
-        """
-        if len(servers) == 0:
-            if clear_others:
-                return self.delete_servers()
-            else:
-                return Effect(Constant(None))
-        query = ('INSERT INTO {cf} ("tenantId", "groupId", last_update, '
-                 'server_id, server_blob, server_as_active) '
-                 'VALUES(:tenantId, :groupId, :last_update, :server_id{i}, '
-                 ':server_blob{i}, :server_as_active{i});')
-        params = merge(self.params, {"last_update": last_update})
-        queries = []
-        for i, server in enumerate(servers):
-            params['server_id{}'.format(i)] = server['id']
-            params['server_as_active{}'.format(i)] = server.pop(
-                '_is_as_active', False)
-            params['server_blob{}'.format(i)] = json.dumps(server)
-            queries.append(query.format(cf=self.table, i=i))
-        if clear_others:
-            return self.delete_servers().on(
-                lambda _: cql_eff(
-                    batch(queries, get_client_ts(self.clock)), params))
-        else:
-            return cql_eff(batch(queries, get_client_ts(self.clock)), params)
+    @do
+    def update_servers(self, time, servers):
+        """
+        See :method:`IScalingGroupServersCache.update_servers`
+
+        This implementation makes two important assumptions which needs to be
+        taken care by caller:
+        - It is not re-entrant for a given object; i.e a call to this must
+          be returned before another can be made
+        - `time` must be higher for subsequent calls.
+        """
+        # get current servers
+        current, last_update = yield self.get_servers(False)
+        if last_update is not None and time <= last_update:
+            raise ValueError(
+                "Given time arg {} must be greater than time of earlier "
+                "inserted servers {}".format(time, last_update))
+
+        # Insert new ones
+        if servers:
+            query = ('INSERT INTO {cf} ("tenantId", "groupId", last_update, '
+                     'server_id, server_blob, server_as_active) '
+                     'VALUES(:tenantId, :groupId, :last_update, :server_id{i},'
+                     ' :server_blob{i}, :server_as_active{i});')
+            params = assoc(self.params, "last_update", time)
+            queries = []
+            for i, server in enumerate(servers):
+                params['server_id{}'.format(i)] = server['id']
```

Comment:
multi-key updates like this I prefer to spell differently, i.e.:

```python
                params.update({
                    'server_id{}'.format(i): server['id'],
                    'server_as_active{}'.format(i): server.pop('_is_as_active', False)
                    'server_blob{}'.format(i): json.dumps(server)
                })
```

If you want to get really fancy, [`merge`](http://toolz.readthedocs.io/en/latest/api.html#toolz.dicttoolz.merge) will let you do this and return a new value, which will also let you avoid mutating `queries` as you build it; `cql_eff(batch([<big generator expression>]))` would be a point-free way of expressing this which means it's a lot easier to see there are no side-effects.